ext/native/gfx_es2/draw_text_android.cpp

```diff
@@ -0,0 +1,302 @@
+}
+
+bool TextDrawerAndroid::IsReady() const {
+	return cls_textRenderer != nullptr && method_measureText != nullptr && method_renderText != nullptr;
+}
+
+uint32_t TextDrawerAndroid::SetFont(const char *fontName, int size, int flags) {
+	// We will only use the default font but just for consistency let's still involve
+	// the font name.
+	uint32_t fontHash = hash::Fletcher((const uint8_t *)fontName, strlen(fontName));
+	fontHash ^= size;
+	fontHash ^= flags << 10;
+
+	auto iter = fontMap_.find(fontHash);
+	if (iter != fontMap_.end()) {
+		fontHash_ = fontHash;
+		return fontHash;
+	}
+
+	// Just chose a factor that looks good, don't know what unit size is in anyway.
+	AndroidFontEntry entry;
+	entry.size = (float)(size * 1.4f) / dpiScale_;
+	fontMap_[fontHash] = entry;
+	fontHash_ = fontHash;
+	return fontHash;
+}
+
+void TextDrawerAndroid::SetFont(uint32_t fontHandle) {
+	uint32_t fontHash = fontHandle;
+	auto iter = fontMap_.find(fontHash);
+	if (iter != fontMap_.end()) {
+		fontHash_ = fontHandle;
+	} else {
+		ELOG("Invalid font handle %08x", fontHandle);
+	}
+}
+
+std::string TextDrawerAndroid::NormalizeString(std::string str) {
+	return ReplaceAll(str, "&&", "&");
+}
+
+void TextDrawerAndroid::MeasureString(const char *str, size_t len, float *w, float *h) {
+	uint32_t stringHash = hash::Fletcher((const uint8_t *)str, len);
+	uint32_t entryHash = stringHash ^ fontHash_;
+
+	TextMeasureEntry *entry;
+	auto iter = sizeCache_.find(entryHash);
+	if (iter != sizeCache_.end()) {
+		entry = iter->second.get();
+	} else {
+		float scaledSize = 14;
+		auto iter = fontMap_.find(fontHash_);
+		if (iter != fontMap_.end()) {
+			scaledSize = iter->second.size;
+		} else {
+			ELOG("Missing font");
+		}
+		std::string text(NormalizeString(std::string(str, len)));
+		jstring jstr = env_->NewStringUTF(text.c_str());
+		uint32_t size = env_->CallStaticIntMethod(cls_textRenderer, method_measureText, jstr, scaledSize);
+		env_->DeleteLocalRef(jstr);
+
+		entry = new TextMeasureEntry();
+		entry->width = (size >> 16);
+		entry->height = (size & 0xFFFF);
+		sizeCache_[entryHash] = std::unique_ptr<TextMeasureEntry>(entry);
+	}
+	entry->lastUsedFrame = frameCount_;
+	*w = entry->width * fontScaleX_ * dpiScale_;
+	*h = entry->height * fontScaleY_ * dpiScale_;
+}
+
+void TextDrawerAndroid::MeasureStringRect(const char *str, size_t len, const Bounds &bounds, float *w, float *h, int align) {
+	double scaledSize = 14;
+	auto iter = fontMap_.find(fontHash_);
+	if (iter != fontMap_.end()) {
+		scaledSize = iter->second.size;
+	} else {
+		ELOG("Missing font");
+	}
+
+	std::string toMeasure = std::string(str, len);
+	if (align & FLAG_WRAP_TEXT) {
+		bool rotated = (align & (ROTATE_90DEG_LEFT | ROTATE_90DEG_RIGHT)) != 0;
+		WrapString(toMeasure, toMeasure.c_str(), rotated ? bounds.h : bounds.w);
+	}
+
+	std::vector<std::string> lines;
+	SplitString(toMeasure, '\n', lines);
+	float total_w = 0.0f;
+	float total_h = 0.0f;
+	for (size_t i = 0; i < lines.size(); i++) {
```

Comment:
Probably good to be consistent, but iirc, the reason for this in Windows was:

> GetTextExtentPoint32 doesn't consider "\n" (new line) or "\r\n" (carriage return and new line) characters when it computes the height of a text string.

-[Unknown]